File: Week 6/1. ROS Launch/ros2_ws/install/turtle_controller/lib/turtle_controller/move_linear.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from turtlesim.msg import Pose
from geometry_msgs.msg import Twist

class MoveLinear(Node):
    def __init__(self):
        super().__init__("move_turtle_linearly")
        self.publisher_ = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.subsciber_ = self.create_subscription(Pose, "/turtle1/pose", self.move_turtle, 10)
        self.get_logger().info("Move turtle by specifying distance, speed and direction")

        self.time = self.get_clock().now().seconds_nanoseconds()
        self.time_sec = self.time[0]

        self.distance = float(input("Turtle should move at distance: "))
        # self.speed = self.distance/self.time_sec

        self.f_b = int(input("Enter [1] - forward, [0] - backward: "))

    def move_turtle(self, pose: Pose):
        origin = 5.5

        cmd = Twist()

        if self.f_b == 1 and pose.x < origin+self.distance:
            cmd.linear.x = self.distance 
        elif self.f_b == 0 and pose.x > origin-self.distance:
            cmd.linear.x = -self.distance

        # if pose.x > origin+self.distance:
        #     cmd.linear.x = 0.0 
           

        self.publisher_.publish(cmd)

def main(args=None):
    rclpy.init(args=args)
    move = MoveLinear()
    rclpy.spin(move)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Week 6/1. ROS Launch/ros2_ws/src/turtle_controller/src/move_linear.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

using std::placeholders::_1;

class MoveLinear : public rclcpp::Node
{
public:
    MoveLinear() : Node("move_turtle_linearly")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        subscriber_ = this->create_subscription<turtlesim::msg::Pose>("/turtle1/pose", 10, std::bind(&MoveLinear::publish_vel, this, _1));
        RCLCPP_INFO(this->get_logger(), "Ready to move the turtle linearly");
        std::cin >> name;
        RCLCPP_INFO(this->get_logger(), "Robot name is: %s", name.c_str());
    }

private:
    void publish_vel(const turtlesim::msg::Pose::SharedPtr pose)
    {
        
        if (pose->x < 1.0 || pose->x > 10.0){
            speed.linear.x = 2.0;
            speed.angular.z = 1.8;
        } else {
            speed.angular.z = 0.0;
            speed.linear.x = 2.0;
        }
        publisher_->publish(speed);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscriber_;
    rclcpp::TimerBase::SharedPtr timer_;
    geometry_msgs::msg::Twist speed;
    std::string name;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MoveLinear>());
    rclcpp::shutdown();
    return 0;
}